debezium-connector-cassandra/src/main/java/io/debezium/connector/cassandra/SnapshotProcessor.java

```diff
@@ -0,0 +1,320 @@
+/*
+ * Copyright Debezium Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.debezium.connector.cassandra;
+
+import com.datastax.driver.core.ColumnMetadata;
+import com.datastax.driver.core.DataType;
+import com.datastax.driver.core.ResultSet;
+import com.datastax.driver.core.Row;
+import com.datastax.driver.core.TableMetadata;
+import com.datastax.driver.core.querybuilder.BuiltStatement;
+import com.datastax.driver.core.querybuilder.QueryBuilder;
+import com.datastax.driver.core.querybuilder.Select;
+import io.debezium.connector.cassandra.exceptions.CassandraConnectorTaskException;
+import io.debezium.connector.cassandra.transforms.CassandraTypeDeserializer;
+import org.apache.avro.Schema;
+import org.apache.cassandra.config.DatabaseDescriptor;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.time.Duration;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.stream.Collectors;
+
+
+/**
+ * This reader is responsible for initial bootstrapping of a table,
+ * which entails converting each row into a change event and enqueueing
+ * that event to the {@link BlockingEventQueue}.
+ *
+ * IMPORTANT: Currently, only when a snapshot is completed will the OffsetWriter
+ * record the table in the offset.properties file (with filename "" and position
+ * -1). This means if the SnapshotProcessor is terminated midway, upon restart
+ * it will skip all the tables that are already recorded in offset.properties
+ */
+public class SnapshotProcessor extends AbstractProcessor {
```

Comment:
How does the transition from snapshot to log reading work? Is it possible to process the logs from the exact position of the snapshot? Or would it more be an approximation?

Reply:
To make it process the logs from exact position of the snapshot is going to be extremely difficult to implement because:
(1) events are logged out of order on a single node
(2) even if we can somehow sort the events on each node, the same event replicated in the cluster will arrive in Kafka in an indeterministic order.

This is why we were not concerned with snapshotting must happen prior to commit log processing. We have both of them starting in separate threads at the same time instead. And since Cassandra is LWW, the 'sorting' is done downstream.

Reply:
Gotcha; what is "LWW"?

Reply:
sorry, last-write-win